### PR TITLE
feat(rpc): websocket fix, use config for allow-list, tests

### PR DESCRIPTION
**Pull request by transmissions12**
Changes -
- changed order for websocket connection set up
- ws tests
- param for ws originAllowlist

### CODE REVIEW

**Comment by transmissions12** [comment]
@itsdevbear still working on sub heads and adding testing scenario for ws

**Comment by itsdevbear** [comment]
🔥🔥🔥

File: cosmos/testing/jsonrpc/eth_test.go [hunk 1]
```diff
@@ -24,11 +24,13 @@ import (
...
 	"math/big"
 	"os"
+	"strings"
 	"testing"
 	"time"
 
-	"github.com/ethereum/go-ethereum"
+	geth "github.com/ethereum/go-ethereum"
 	"github.com/ethereum/go-ethereum/accounts/abi/bind"
+	gethtypes "github.com/ethereum/go-ethereum/core/types"
```
Comment (itsdevbear) [inline cosmos/testing/jsonrpc/eth_test.go:33]: I'd put the websocket tests in a different file tbh.

File: eth/go.mod [hunk 2]
```diff
@@ -3,7 +3,7 @@ module pkg.berachain.dev/polaris/eth
 go 1.20
 
 // Required for stateful precompiles and supporting the Ethereum JSON-RPC API.
-replace github.com/ethereum/go-ethereum => github.com/berachain/polaris-geth v0.0.0-20230313222742-b18b15fecd70
+replace github.com/ethereum/go-ethereum => github.com/berachain/polaris-geth v0.0.0-20230321144626-29251e594538
```
Comment (calbera) [inline eth/go.mod:6]: you can add this to the cosmos/ and playground/ go.mod's as well

**Comment by calbera** [comment]
When do we want to emit new txs on the subscription for new txs? `SubscribeNewTxsEvent()` Because geth weirdly only sends txs to this feed on reorgs https://github.com/berachain/polaris-geth/blob/e3ca3ff6bc37871160f3e1bcd9e992c660373010/core/txpool/txpool.go#L1288. One option is to send to the feed when a tx gets added to the mempool. The other option that makes sense to me is send to the feed once the tx is committed to comet and is being processed in processTransaction. @itsdevbear @transmissions12

**Comment by itsdevbear** [comment]
You sure? That seems kinda cooked LOL

**Comment by itsdevbear** [comment]
I'd just emit an event when it gets inserted to the app side.

**Comment by calbera** [comment]
true its weird, the question becomes do you want a new tx event for smth that may not be finalized?

**Comment by transmissions12** [comment]
> Member

I think it boils down to purpose of that subscribe. If purpose is to know transaction as soon as it enters (no consensus), we should do it at the start. I would say it's at the start (without finalize) because you have subscribe blocks that finalize transactions.

File: eth/rpc/config/config.go [hunk 3]
```diff
@@ -43,5 +43,8 @@ type (
 		// RPCTxFeeCap is the global transaction fee(price * gaslimit) cap for
 		// send-transaction variants. The unit is ether.
 		RPCTxFeeCap float64 `mapstructure:"rpc-tx-fee-cap"`
+
+		// Allowlist defines a list of allowed origins for the JSON-RPC server.
+		Allowlist []string `mapstructure:"allowlist"`
```
Comment (itsdevbear) [inline eth/rpc/config/config.go:48]: OriginAllowlist?

File: go.work [hunk 4]
```diff
@@ -8,3 +8,5 @@ use (
 	./magefiles
 	./playground
 )
+
```
Comment (itsdevbear) [inline go.work:11]: we can probably remove this from here?

File: magefiles/cosmos.go [hunk 5]
```diff
@@ -162,5 +162,5 @@ func (c Cosmos) TestUnit() error {
 // Runs all integration for the Cosmos SDK chain.
 func (c Cosmos) TestIntegration() error {
 	LogGreen("Running integration tests for the Cosmos SDK chain.")
-	return testIntegration(c.directory())
+	return testIntegration(c.directory() + "/testing")
```
Comment (itsdevbear) [inline magefiles/cosmos.go:165]: bullish

File: .gitignore [hunk 6]
```diff
@@ -191,6 +191,7 @@ pkg/**/.*.yaml
 /tmp
 
 # forge
+eth/testutil/contracts
```
Comment (itsdevbear) [inline .gitignore:194]: remove this

File: cosmos/rpc/register.go [hunk 7]
```diff
@@ -30,7 +30,7 @@ import (
 // with routes on the native Cosmos API Server.
 func RegisterJSONRPCServer(ctx client.Context, rtr *mux.Router, provider Provider) {
 	cfg := provider.GetConfig()
+	rtr.PathPrefix(cfg.BaseRoute + "/ws").Handler(provider.GetWS().WebsocketHandler(cfg.Allowlist))
```
Comment (itsdevbear) [inline cosmos/rpc/register.go:33]: bullish

File: cosmos/testing/integration/rpc/eth_test.go [hunk 8]
```diff
@@ -45,76 +49,169 @@ func TestRpc(t *testing.T) {
 
 var _ = Describe("Network", func() {
 	var client *ethclient.Client
+	var wsClient *ethclient.Client
```
Comment (itsdevbear) [inline cosmos/testing/integration/rpc/eth_test.go:52]: lgtm, but need to fix everything up as you said before.

**Review by itsdevbear — APPROVED** [review]
lgtm, just the few things we discussed for later PRs

File: eth/go.mod (hunk 2 above)
Comment (transmissions12) [reply eth/go.mod]: done

**Comment by itsdevbear** [comment]
@transmissions12 if u need help with getting this one closed out lmk and i can crush it tomorrow with u

**Comment by transmissions12** [comment]
let's crush it in the AM💪

**Comment by itsdevbear** [comment]
🫡🫡🫡🫡🫡

image

File: go.work [hunk 9]
```diff
@@ -8,3 +8,4 @@ use (
 	./magefiles
 	./playground
 )
+replace github.com/ethereum/go-ethereum => github.com/berachain/polaris-geth v0.0.0-20230321213944-019945fbf8fc
```
Comment (itsdevbear) [inline go.work:11]: remove

File: eth/rpc/config/default.go [hunk 10]
```diff
@@ -41,13 +41,17 @@ const (
 	DefaultRPCEVMTimeout = 5 * time.Second
 )
 
+// DefaultOriginAllowlist is the default origin allowlist (allows all by default) for the JSON-RPC server.
+var DefaultOriginAllowlist = []string{"*"}
```
[the remark this answers is not in the document]
Comment (transmissions12) [reply eth/rpc/config/default.go:45]: done

File: go.work (hunk 9 above)
Comment (transmissions12) [reply go.work:11]: done

File: cosmos/testing/integration/rpc/ws_test.go [hunk 11]
```diff
@@ -0,0 +1,80 @@
...
+)
+
+var _ = Describe("Network with WS", func() {
+	var ctx context.Context
+	BeforeEach(func() {
+		ctx = context.Background()
+	})
+	It("should connect -- multiple ", func() {
+		// Dial an Ethereum websocket Endpoint
+		wsaddr := "ws:" + strings.TrimPrefix(tf.Network.Validators[0].APIAddress+"/eth/rpc/ws/", "http:")
```
Comment (itsdevbear) [inline cosmos/testing/integration/rpc/ws_test.go:45]: upgrade to use the Fixtures.

Comment (transmissions12) [reply cosmos/testing/integration/rpc/ws_test.go:45]: this is checking if we can create multiple ws connections, already using fixtures

File: cosmos/testing/integration/rpc/eth_test.go (hunk 8 above)
Comment (transmissions12) [reply cosmos/testing/integration/rpc/eth_test.go]: bless

File: cosmos/testing/integration/rpc/eth_test.go [hunk 12]
```diff
@@ -59,81 +65,100 @@ var _ = SynchronizedAfterSuite(func() {
 })
 
 var _ = Describe("Network", func() {
-	It("should support eth_chainId", func() {
-		chainID, err := tf.EthClient.ChainID(context.Background())
-		Expect(err).ToNot(HaveOccurred())
-		Expect(chainID.String()).To(Equal("69420"))
+	var ctx context.Context
```
[the remark this answers is not in the document]
Comment (transmissions12) [reply cosmos/testing/integration/rpc/eth_test.go:68]: oops had taken them out, re-added em

**Comment by itsdevbear** [comment]
@transmissions12 integration test fails but other than that lgtm

**Comment by itsdevbear** [comment]
cloudflare is also cooked idk why

**Comment by transmissions12** [comment]
the integration test is an intermittent failure. We don't have it all the time, reason being the `BuildTransactor` utils fn. Have a look at the logs and let me know how we should proceed @itsdevbear

**Review by itsdevbear — APPROVED** [review]

**Pull request merged**